zebra-chain/src/addresses/sapling.rs

```diff
@@ -1 +1,187 @@
+    diversifier: sapling::Diversifier,
+    transmission_key: sapling::TransmissionKey,
+}
+
+impl fmt::Debug for SaplingShieldedAddress {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        f.debug_struct("SaplingShieldedAddress")
+            .field("diversifier", &self.diversifier)
+            .field("transmission_key", &self.transmission_key)
+            .finish()
+    }
+}
+
+impl SaplingShieldedAddress {
+    fn to_human_readable_address(&self, network: Network) -> Result<String, bech32::Error> {
+        let mut bytes = io::Cursor::new(Vec::new());
+        let _ = self.zcash_serialize(&mut bytes);
+
+        let mut hrp = "";
+
+        match network {
+            Network::Mainnet => hrp = human_readable_parts::MAINNET,
+            _ => hrp = human_readable_parts::TESTNET,
+        }
+
+        bech32::encode(hrp, bytes.get_ref().to_base32())
+    }
+}
+
+impl Eq for SaplingShieldedAddress {}
+
+impl ZcashSerialize for SaplingShieldedAddress {
+    fn zcash_serialize<W: io::Write>(&self, mut writer: W) -> Result<(), io::Error> {
+        writer.write_all(&self.diversifier.0[..])?;
+        writer.write_all(&self.transmission_key.to_bytes())?;
+
+        Ok(())
+    }
+}
+
+impl ZcashDeserialize for SaplingShieldedAddress {
+    fn zcash_deserialize<R: io::Read>(mut reader: R) -> Result<Self, SerializationError> {
+        let mut diversifier_bytes = [0; 11];
+        reader.read_exact(&mut diversifier_bytes)?;
+
+        let transmission_key_bytes = reader.read_32_bytes()?;
+
+        Ok(SaplingShieldedAddress {
+            diversifier: sapling::Diversifier(diversifier_bytes),
+            transmission_key: sapling::TransmissionKey::from_bytes(transmission_key_bytes),
+        })
+    }
+}
+
+impl std::str::FromStr for SaplingShieldedAddress {
+    type Err = SerializationError;
+
+    fn from_str(s: &str) -> Result<Self, Self::Err> {
+        match bech32::decode(s) {
+            Ok((_, bytes)) => {
+                let decoded = Vec::<u8>::from_base32(&bytes).unwrap();
+                Self::zcash_deserialize(io::Cursor::new(decoded))
+            }
+            Err(_) => Err(SerializationError::Parse("bech32 decoding error")),
+        }
+    }
+}
+
+#[cfg(test)]
+impl Arbitrary for SaplingShieldedAddress {
+    type Parameters = ();
+
+    fn arbitrary_with(_args: Self::Parameters) -> Self::Strategy {
+        (array::uniform11(any::<u8>()), array::uniform32(any::<u8>()))
+            .prop_map(|(diversifier_bytes, transmission_key_bytes)| {
+                return Self {
+                    diversifier: sapling::Diversifier(diversifier_bytes),
+                    transmission_key: sapling::TransmissionKey::from_bytes(transmission_key_bytes),
+                };
+            })
+            .boxed()
+    }
+
+    type Strategy = BoxedStrategy<Self>;
+}
+
+#[cfg(test)]
+mod tests {
+
+    use rand_core::OsRng;
+    use std::str::FromStr;
+
+    use super::*;
+
+    #[test]
+    fn from_str_debug() {
```

Comment:
Do we want to encode the Debug format into tests, or keep it implementation-defined?

Reply:
We don't have to, I think this was just used to make sure it was behaving correctly, removing.